provider/tencentcloud/dnspod.go

```diff
@@ -0,0 +1,263 @@
+	createEndpoints := make(map[string][]*endpoint.Endpoint)
+	for zoneId, _ := range zoneNameIDMapper {
+		createEndpoints[zoneId] = make([]*endpoint.Endpoint, 0)
+	}
+	for _, change := range [][]*endpoint.Endpoint{changes.Create, changes.UpdateNew} {
+		for _, createChange := range change {
+			if zoneId, _ := zoneNameIDMapper.FindZone(createChange.DNSName); zoneId != "" {
+				createEndpoints[zoneId] = append(createEndpoints[zoneId], createChange)
+			}
+		}
+	}
+	if err := p.createRecord(recordsGroupMap, createEndpoints); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (p *TencentCloudProvider) createRecord(zoneMap map[uint64]*RecordListGroup, endpointsMap map[string][]*endpoint.Endpoint) error {
+	for zoneId, endpoints := range endpointsMap {
+		zoneIdString, _ := strconv.ParseUint(zoneId, 10, 64)
+		domain := zoneMap[zoneIdString]
+		for _, endpoint := range endpoints {
+			for _, target := range endpoint.Targets {
+				if endpoint.RecordType == "TXT" && strings.HasPrefix(target, "\"heritage=") {
+					target = strings.Trim(target, "\"")
+				}
+				if err := p.createRecords(domain.Domain, endpoint, target); err != nil {
+					return err
+				}
+			}
+		}
+	}
+	return nil
+}
+
+func (p *TencentCloudProvider) createRecords(domain *dnspod.DomainListItem, endpoint *endpoint.Endpoint, target string) error {
+	request := dnspod.NewCreateRecordRequest()
+
+	request.Domain = common.StringPtr(*domain.Name)
+	request.RecordType = common.StringPtr(endpoint.RecordType)
+	request.Value = common.StringPtr(target)
+	request.SubDomain = common.StringPtr(getSubDomain(*domain.Name, endpoint))
+	if endpoint.RecordTTL.IsConfigured() {
+		request.TTL = common.Uint64Ptr(uint64(endpoint.RecordTTL))
+	}
+	request.RecordLine = common.StringPtr("默认")
+
+	if _, err := p.apiService.CreateRecord(request); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (p *TencentCloudProvider) deleteRecords(RecordIdsMap map[string][]uint64) error {
+	for domain, recordIds := range RecordIdsMap {
+		if len(recordIds) == 0 {
+			continue
+		}
+		if err := p.deleteRecord(domain, recordIds); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (p *TencentCloudProvider) deleteRecord(domain string, recordIds []uint64) error {
+	request := dnspod.NewDeleteRecordRequest()
+	request.Domain = common.StringPtr(domain)
+
+	for _, recordId := range recordIds {
+		request.RecordId = common.Uint64Ptr(recordId)
+		if _, err := p.apiService.DeleteRecord(request); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func groupDomainRecordList(recordListGroup map[uint64]*RecordListGroup) (endpointMap map[string]*RecordListGroup) {
+	endpointMap = make(map[string]*RecordListGroup)
+
+	for _, recordGroup := range recordListGroup {
+		for _, record := range recordGroup.RecordList {
+			key := *record.Type + ":" + *record.Name + "." + *recordGroup.Domain.Name
+			if *record.Name == TencentCloudEmptyPrefix {
+				key = *record.Type + ":" + *recordGroup.Domain.Name
```

Comment:
I think you don't need to dereference pointers here (ibmcloud seems to also does it, but all others not) and basically all over the place. See also the inmemory provider.

Reply:
Sorry, I just not understand only this feedback. I wants to group the target by dnsname and type here.

Reply:
Ok sorry!
You can

Reply:
key := record.Type + ":" + record.Name + "." + recordGroup.Domain.Name

Reply:
Yes like that and in general I prefer fmt.Sprintf for string concatenation.

Reply:
record.Type is *string type. remove the "*" make go language syntax error.